syntax = "proto3";

package node;

service NodeService {
    rpc JoinCluster(JoinClusterRequest) returns (JoinClusterResponse);
    rpc StoreData(StoreDataRequest) returns (StoreDataResponse);
    rpc RetrieveData(RetrieveDataRequest) returns (RetrieveDataResponse);
    rpc GetConnectedNodeIP(GetConnectedNodeIPRequest) returns (GetConnectedNodeIPResponse);
    rpc NotifyNodeJoined(NotifyNodeJoinedRequest) returns (NotifyNodeJoinedResponse);
}

message JoinClusterRequest {
    string node_id = 1;
    string node_name = 2;
    string ip_address = 3;
}

message JoinClusterResponse {}

message StoreDataRequest {
    string key = 1;
    string value = 2;
}

message StoreDataResponse {}

message RetrieveDataRequest {
    string key = 1;
}

message RetrieveDataResponse {
    string value = 1;
}

message GetConnectedNodeIPRequest {
    string node_id = 1;
}

message GetConnectedNodeIPResponse {
    string ip_address = 1;
}

message NotifyNodeJoinedRequest {
    string node_id = 1;
    string node_name = 2;
    string ip_address = 3;
}

message NotifyNodeJoinedResponse {}